state/src/skip_list.rs

```diff
@@ -0,0 +1,913 @@
+                trie.insert(fv_key.key_hash().as_slice(), &end_of_list)?;
+            }
+        }
+        // For levels after the reinsert level, we must delete the old key and the new
+        // head points where the old head used to point (or to the end of the list if
+        // there was no previous reference).
+        for level in (reinsert_level + 1)..=max_levels {
+            let fv_key = SkipListKey::new(account, level, first_value.as_ref());
+            let insert_key = SkipListKey::new(account, level, element);
+
+            let fv_trie_key = fv_key.key_hash();
+            let fv_value = SkipListValue::<T>::read_trie(&fv_trie_key, trie)?;
+            match fv_value {
+                Some(value) => {
+                    trie.remove(fv_trie_key.as_slice())?;
+                    trie.insert(insert_key.key_hash().as_slice(), &value.serialize())?;
+                }
+                None => trie.insert(insert_key.key_hash().as_slice(), &end_of_list)?,
+            }
+        }
+
+        return Ok(());
+    }
+
+    // For each height the new key occupies, insert it into the corresponding
+    // linked list.
+    let insert_level = pick_insert_level(rng);
+    let insert_value = SkipListValue::<T> {
+        next_value: Some(Cow::Borrowed(element)),
+    }
+    .serialize();
+    let prevs = collect_predecessors(
+        account,
+        head.current_highest_level,
+        first_value.clone(),
+        element,
+        trie,
+    )?;
+    let mut updated_heights = 0;
+    for (key, original_value) in prevs.into_iter().flatten().take(insert_level + 1) {
+        trie.insert(key.key_hash().as_slice(), &insert_value)?;
+
+        let insert_key = SkipListKey::new(account, key.level, element);
+        trie.insert(
+            insert_key.key_hash().as_slice(),
+            &original_value.serialize(),
+        )?;
+
+        updated_heights += 1;
+    }
+
+    if updated_heights <= insert_level {
+        // New element is taller than previous max height.
+        // Add new keys pointing from first to new element and update head value.
+        let mut level = head.current_highest_level + 1;
+        while updated_heights <= insert_level {
+            let key = SkipListKey::new(account, level, first_value.as_ref());
+            trie.insert(key.key_hash().as_slice(), &insert_value)?;
+
+            let insert_key = SkipListKey::new(account, level, element);
+            trie.insert(insert_key.key_hash().as_slice(), &end_of_list)?;
+
+            level += 1;
+            updated_heights += 1;
+        }
+
+        let new_head = SkipListHeadValue::new(level - 1, first_value.as_ref());
+        trie.insert(trie_head_key.as_slice(), &new_head.serialize())?;
+    }
+
+    Ok(())
+}
+
+/// Each type that implements `Listable` must have a unique value for `kind`.
+/// It is used as a discriminator to make the byte-level representation of the keys
+/// in the trie unique.
+pub trait Listable {
+    fn kind() -> u8;
```

Comment:
Is this used part of the serialization in the `element`? There won't be any overlap between modules and resources, right, when inserted in the list?

Reply:
Yes, checked and the key hashes indeed change with different `kind`s, so that differentiates the types.